src/hardware/vga_draw.cpp

```diff
@@ -1265,26 +1307,44 @@ void VGA_CheckScanLength(void) {
 	}
 }
 
-void VGA_ActivateHardwareCursor(void) {
-	bool hwcursor_active=false;
-	if (svga.hardware_cursor_active) {
-		if (svga.hardware_cursor_active()) hwcursor_active=true;
-	}
-	if (hwcursor_active) {
-		switch(vga.mode) {
-		case M_LIN32:
-			VGA_DrawLine=VGA_Draw_LIN32_Line_HWMouse;
-			break;
-		case M_LIN15:
-		case M_LIN16:
-			VGA_DrawLine=VGA_Draw_LIN16_Line_HWMouse;
-			break;
-		default:
-			VGA_DrawLine=VGA_Draw_VGA_Line_HWMouse;
+// Update VGA line drawing function pointer based on the hardware cursor state
+// and current mode's bit-depth. Returns the bits per line-pixel (typically 8
+// bpp)
+uint8_t VGA_ActivateHardwareCursor()
+{
+	uint8_t bit_per_line_pixel = 8;
+
+	auto should_draw_with_hw_mouse = [&]() {
+		return svga.hardware_cursor_active && svga.hardware_cursor_active();
+	};
+	switch (vga.mode) {
+	case M_LIN8:
+		if (should_draw_with_hw_mouse()) {
+			VGA_DrawLine = VGA_Draw_VGA_Line_HWMouse;
+		} else {
+			// Otherwise let 8-bit VESA modes draw from the DAC palette
```

Comment:
Maybe it's me, but I had to read this a few times and it's still ambiguous a bit. Ultimately the code made it clear what we're doing. "Drawing from the DAC palette"--we always do that, the question is how 😄 

Maybe say something like this: "Otherwise we render 8-bit VESA modes as 32-bit RGB (XRGB8888) instead of paletted." Then a further explanation what this fixes perhaphs.

Reply:
Added more comments in this switch statement to contrast how the high and true-colour modes are handled versus the 8-bit mode.

In @Grounded0 's diagram, it shows true and high-colour modes' buffers bypassing the 8-bit->18-bit RGB666 palette LUT, and instead their buffers go straight out to the 8/6 bit DAC for rendering.

Recall `machine = vgadraw`: for the standard EGA and VGA modes, it takes each 8-bit pixel value and looks up the 18-bit RGB666 palette LUT value (well, previously it was a 16-bit RGB565 LUT) on a per-line basis.

After we combined `vgadraw`, it turns out 8-bit VESA mode wasn't using the 18-bit palette LUT (instead, it was using `VGA_Draw_Linear_Line`, which is the same routine used by the true and high colour modes that bypass the 16-bit palette LUT). 

So this PR just adds case to handle do the 8-bit index -> 18-bit RGB666 palette LUT for 8-bit VESA:

![2023-05-17_18-36](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/475fa6b1-513d-4c51-92f2-1e8bcf12bec2)

And the route that the palette gets adjusted (Send Color is called from DAC's IO write ports functions):

![2023-05-17_18-23](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/7328a516-a89a-482b-8079-922edfe5b531)

> How does this fix the 7th Guest problem then

It uses 8-bit VESA and makes small adjustments to the DAC palette (not mid-frame, but either prior or after drawing), and only every now and then.  So the fix is to actually lookup values from there.

> Specifically, is this a proper fix for a well-understood emulation problem

Yes - it should handle any game that's using 8-bit VESA and making DAC palette tweaks.

> how can we test the VESA hardware cursor stuff for regressions?

That code path is the same: if hardware cursor is needed, it's using `VGA_Draw_VGA_Line_HWMouse` which is what it previous used. 

But this function /doesn't/ perform palette lookups. However, I'd rather wait for evidence of games needing this before figuring out how to do it.. because it's quite an involved function.